### PR TITLE
py-nptyping, py-typish: add package

### PR DESCRIPTION
**Pull request by heerener**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-nptyping/package.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from spack.package import *
+
+
+class PyNptyping(PythonPackage):
+    """Type hints for numpy"""
+    homepage = "https://github.com/ramonhagenaars/nptyping"
+    pypi = "nptyping/nptyping-2.4.1.tar.gz"
+
+    version("2.4.1", sha256="57ba684ee5fc5eb681ee04270ee94adb879e4372ce6b640defa08ace8e1df295")
+    version(
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nptyping/package.py]: Is there any reason to add these ancient versions that don't have sdists on PyPI?

File: var/spack/repos/builtin/packages/py-nptyping/package.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        url="https://github.com/ramonhagenaars/nptyping/archive/v1.4.1.tar.gz",
+    )
+    version(
+        "1.0.1",
+        sha256="a00e672bfdaddc99aa6b25dd1ae89d7d58d2b76e8ad099bd69577bac2598589f",
+        url="https://github.com/ramonhagenaars/nptyping/archive/v1.0.1.tar.gz",
+    )
+
+    depends_on("py-setuptools", type="build")
+    depends_on("python@3.5:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nptyping/package.py]: 3.7 in the latest release. But Spack only supports 3.7+, so you could just remove this line

File: var/spack/repos/builtin/packages/py-nptyping/package.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyNptyping(PythonPackage):
+    """Type hints for numpy"""
+    homepage = "https://github.com/ramonhagenaars/nptyping"
+    pypi = "nptyping/nptyping-2.4.1.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nptyping/package.py]: The PyPI tarball does not seem to include the `requirements.txt` file, does this package even build?

File: var/spack/repos/builtin/packages/py-nptyping/package.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyNptyping(PythonPackage):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nptyping/package.py:9]: This package seems to be unrelated to py-typish, was this accidentally included in the wrong PR?

File: var/spack/repos/builtin/packages/py-typish/package.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+from spack.package import *
+
+
+class PyTypish(PythonPackage):
+    """Python library for type checking
+
+    * Functions for thorough checks on types
+    * Instance checks considering generics
+    * Typesafe Duck-typing
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-typish/package.py]: `spack info py-typish` removes newline characters so the formatting of this will be messed up. I would either simplify the docstring or add commas/semi-colons.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-nptyping/package.py (hunk 4 above)
Comment (heerener) [reply var/spack/repos/builtin/packages/py-nptyping/package.py:9]: I've included it here because it depends on py-typish. Will update the PR title to reflect this.

File: var/spack/repos/builtin/packages/py-nptyping/package.py (hunk 1 above)
Comment (heerener) [reply var/spack/repos/builtin/packages/py-nptyping/package.py]: I'm mostly upstreaming changes that have been lying around here for a long time, some of these ancient versions are, apparently, still being used here. I've added the latest version specifically for this PR (although I admit I should've tested properly).

On a related note: as I have more that I'd like to upstream, is it better to make a bunch of smaller PRs with packages that relate to one another, or reduce the PR-spam a little by making a bigger one that contains unrelated packages?

File: var/spack/repos/builtin/packages/py-nptyping/package.py (hunk 3 above)
Comment (heerener) [reply var/spack/repos/builtin/packages/py-nptyping/package.py]: There's even [an issue](https://github.com/ramonhagenaars/nptyping/issues/98) about it - I've linked to the github tarball for now (and installed it, so that one does work). Will revisit once the issue is fixed.

File: var/spack/repos/builtin/packages/py-nptyping/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-nptyping/package.py]: We try to follow a one-package-per-PR system, although it's fine to combine 2+ packages if the changes to one package depend on another.

File: var/spack/repos/builtin/packages/py-nptyping/package.py (hunk 3 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-nptyping/package.py]: Can we remove `pypi` and set `url` globally for now? Keep the comment though, that's useful to refer to if it ever gets fixed.

**Review by adamjstewart — APPROVED** [review]
